app/controllers/catalog_controller.rb

```diff
@@ -1859,8 +1884,10 @@ def fetch_ovf_template_details
 
     ovf_template_details[:provisioning][:disk_format] = provision[:disk_format]
 
-    folder = EmsFolder.find_by(:id => provision[:ems_folder_id])
-    ovf_template_details[:provisioning][:ems_folder_name] = folder.try(:name)
+    vm_folder = dc.folders.find { |f| f.name == 'vm' } if dc
+    ovf_template_details[:provisioning][:ems_folder_name] = vm_folder.ems_ref if vm_folder
+    folder = EmsFolder.find_by(:id => provision[:ems_folder_id]) if !dc
+    ovf_template_details[:provisioning][:ems_folder_name] = folder.try(:name) if folder
```

Comment:
The way this is written, an initial read makes it look like both directions are possible, when they aren't - I think this might read a little clearer:

```suggestion
    if dc
      folder = dc.folders.find { |f| f.name == 'vm' }
      ovf_template_details[:provisioning][:ems_folder_name] = vm_folder.ems_ref if folder
    else
      folder = EmsFolder.find_by(:id => provision[:ems_folder_id])
      ovf_template_details[:provisioning][:ems_folder_name] = folder.try(:name) if folder
    end
```

---

Completely separately - this is a very vmware specific rule.  While this method is an ovf related method, which is also vmware specific, IMO none of this belongs in the UI controller.  We need to find a way to get these specifics out of here and into the vmware provider - cc @agrare

Reply:
Completely separately - this is a very vmware specific rule.  While this method is an ovf related method, which is also vmware specific, IMO none of this belongs in the UI controller.  We need to find a way to get these specifics out of here and into the vmware provider - cc @agrare

Reply:
Yeah this is typically handled in the provider plugin during the provision (at least on the VM provision side).  The selected folder object is passed in and the vmware provider handles the "is it a datacenter, then switch to the vm folder"

Reply:
Should the change to move this to the provider side happen in this pr or should this be merged and another pr be made after?

Reply:
I don't think we can realistically do that in this PR, this whole chunk of code looks like it needs refactoring
```
  def class_service_template(prov_type)
    if content_library?
      ManageIQ::Providers::Vmware::InfraManager::OvfServiceTemplate
    elsif prov_type.starts_with?('generic')
      prov_type.gsub(/(generic)(_.*)?/, 'service_template\2').classify.constantize
    else 
      ServiceTemplate
    end  
  end
```

Along with tons of other class checks like `if @record.kind_of?(ServiceTemplateAnsibleTower)` and `if @record.kind_of?(ServiceTemplateContainerTemplate)`

I have to admit I don't understand the generic service provisioning well but there is a lot of provider specific code in this controller.